### PR TITLE
matcher refactor

### LINKED ISSUE #14: extension actions can use throttle manager too
we should prevoide a options liks `throttle:number`, and the extension actions  can use inner throttler manager too.

### LINKED ISSUE #15: store previous step , enhance the step record
we should save the previous action, and in some conditions, we may need to drop the previous action or collaborate the presious action and next actions.
for example:
1. previsou click step + next input file step = drop previous click step + next input file step
2. previous click step + next click step  = next double click step

### PR DESCRIPTION
**Pull request by Sczlog**
use fsm as matcher to match event type.

### CODE REVIEW

File: __tests__/util/mock.ts [hunk 1]
```diff
@@ -1,26 +1,24 @@
...
 import { mocked } from 'ts-jest/utils';
 import { IMatcher } from '../../src/matcher';
-import { IObserver } from '../../src/observers';
+import { AbstractObserver } from '../../src/observers';
 
 const emptyFn = () => void 0;
 
 export const getMockedObserver = () => {
-  return mocked<IObserver>({
+  return mocked<any>({
```
Comment (netweng) [inline __tests__/util/mock.ts:9]: can we remove any?

Comment (Sczlog) [reply __tests__/util/mock.ts:9]: we cannot remove it due to its have some protected function in it and I cannot mock it directly.

**Comment by netweng** [comment]
LGTM

**Comment by netweng** [comment]
has conflict

**Comment by Sczlog** [comment]
> has conflict

conflict resolved

**Pull request merged**